apps/src/blockly/addons/cdoSpritePointer.js

```diff
@@ -0,0 +1,112 @@
+// Helper functions for pointer block logic. In our pointer blocks (a.k.a.)
+// mini toolbox blocks) we "shadow" the image on the block from a source block.
+// This file contains functions to get and set the image on the pointer block.
+
+/**
+ * Get the image url for the given pointer block.
+ * @param {Block} block pointer block to get the image url for
+ * @param {*} pointerMetadataMap Object of the form {blockType: {imageSourceType: <string>, imageIndex: <number>}},
+ *  which maps pointer block types to the image source block type they can shadow and the index of the image to shadow.
+ * @param {string} imageSourceId Optional id of the image source block.
+ * @returns The url of the image that the pointer block should display, or an empty string if the block should display no image
+ */
+export function getPointerBlockImageUrl(
+  block,
+  pointerMetadataMap,
+  imageSourceId
+) {
+  const pointerData = pointerMetadataMap[block.type];
+  if (!pointerData || !block.inputList || block.inputList.length === 0) {
+    return '';
+  }
+  let imageSource = undefined;
+  const mainWorkspace = Blockly.getMainWorkspace();
+  if (imageSourceId !== undefined) {
+    imageSource = mainWorkspace.getBlockById(imageSourceId);
+  }
+  if (!imageSource) {
+    const rootBlock = block.getRootBlock();
+    if (rootBlock.type === pointerData.imageSourceType) {
+      imageSource = rootBlock;
+    } else if (block.getRootBlock().type === block.type) {
+      const blockWorkspace = Blockly.Workspace.getById(block.workspace.id);
+      // If this block has itself as a root and is in a flyout workspace,
```

Comment:
Are we check for block type equality above to determine if the block has itself as a root block? If so, I wonder if `.id` might be more reliable.

Reply:
good call, will update